### PR TITLE
Allow running tests when stdlib is built with LTO by working around auto-linking

### PR DESCRIPTION
**Pull request by kubamracek**

### CODE REVIEW

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

File: test/lit.cfg [hunk 1]
```diff
@@ -877,6 +877,15 @@ if run_vendor == 'apple':
         resource_dir_opt = "-resource-dir %s" % new_resource_dir
         lit_config.note('Using freestanding resource dir: ' + new_resource_dir)
 
+    # Auto-linking does not work when stdlib is built with LTO, because linked
+    # libraries are discovered too late (after optimizations are applied). Let's
+    # explicitly link all libraries needed in tests.
+    if "stdlib_lto" in config.available_features:
```
Comment (compnerd) [inline test/lit.cfg:883]: I think that this really needs a comment explaining that this won't work on Linux (without LLD).  The autolinking mechanism on Linux will fail due to needing `-L` arguments to the path even with this forced linking as the extracted commands will force the linker to look up the linking by name as well.

Comment (kubamracek) [reply test/lit.cfg:883]: Can you expand on this a bit? I thought we're already passing in `-L` to the resource dir. Why is Linux different from Darwin here? Forgive my ignorance, please :)

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
Resurrecting this old PR, because the problem is still present. At least with ld64 (and possibly other linkers too?), auto-linking doesn't play well with LTO. I have requested fixes from the linker, but to unblock progress (towards enabling LTO on the stdlib on the "minimal" build preset) I'd like to propose to go forward with this workaround.

@compnerd Any chance you could once more explain the Linux linking specifics here? For now, I have included your comment verbatim, but I don't really follow tbh 😃

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@compnerd gentle ping :)

File: test/lit.cfg (hunk 1 above)
Comment (compnerd) [reply test/lit.cfg:956]: It isn't the resource dir that is always the location for the library - the resource dir is meant for a small subset of libraries, those which are needed by the compiler directly (i.e. stdlib).  The other libraries can be anywhere.  The autolink only indicates that the library must be linked, not where to find it.  The normal lookup order/path applies for the resolution of linked libraries, even when auto-linked.  LLD will encode the `-l...` just as much as what we do now, so ultimately, the normal linker search rules apply.

**Review by compnerd — APPROVED** [review]
Discussed with @kubamracek offline; with some minor tweaks to the comment, this seems fine.

**Comment by kubamracek** [comment]
@swift-ci please test

**Comment by kubamracek** [comment]
@swift-ci please test Windows platform

**Comment by kubamracek** [comment]
@swift-ci please test macOS platform

**Comment by kubamracek** [comment]
I accidentally clicked "merge" before pushing the comment change, sorry about that 😞 Fixing that here: https://github.com/apple/swift/pull/39362

**Pull request merged**